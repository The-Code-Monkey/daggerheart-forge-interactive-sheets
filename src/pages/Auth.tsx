import { JSX, useState } from "react";
import {
  Card,
  CardContent,
  CardDescription,
  CardHeader,
  CardTitle,
} from "@/components/ui/card";
import { Button } from "@/components/ui/button";
import { Input } from "@/components/ui/input";
import { Label } from "@/components/ui/label";
import { Separator } from "@/components/ui/separator";
import { Sparkles } from "lucide-react";
import { supabase } from "@/integrations/supabase/client";
import { useToast } from "@/hooks/use-toast";
import { useNavigate } from "react-router-dom";
import { Turnstile } from "@marsidev/react-turnstile";

const Auth = (): JSX.Element => {
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [username, setUsername] = useState("");
  const [isLogin, setIsLogin] = useState(true);
  const [loading, setLoading] = useState(false);

  // Honeypot field (should remain empty)
  const [honeypot, setHoneypot] = useState("");

  // Simple math captcha
  const [captchaAnswer, setCaptchaAnswer] = useState("");
  const [captchaQuestion] = useState(() => {
    const num1 = Math.floor(Math.random() * 10) + 1;
    const num2 = Math.floor(Math.random() * 10) + 1;
    return {
      question: `${String(num1)} + ${String(num2)}`,
      answer: num1 + num2,
    };
  });

  // Cloudflare Turnstile
  const [turnstileToken, setTurnstileToken] = useState("");

  const { toast } = useToast();
  const navigate = useNavigate();

  const handleAuth = async (e: React.FormEvent) => {
    e.preventDefault();
    setLoading(true);

    // Check honeypot - if filled, it's likely a bot
    if (honeypot) {
      toast({
        title: "Error",
        description: "Suspicious activity detected. Please try again.",
        variant: "destructive",
      });
      setLoading(false);
      return;
    }

    // Check captcha for registration only
    if (!isLogin && parseInt(captchaAnswer) !== captchaQuestion.answer) {
      toast({
        title: "Error",
        description: "Please solve the math problem correctly.",
        variant: "destructive",
      });
      setLoading(false);
      return;
    }

    // Check Turnstile token for registration
    if (!isLogin && !turnstileToken) {
      toast({
        title: "Error",
        description: "Please complete the security verification.",
        variant: "destructive",
      });
      setLoading(false);
      return;
    }

    try {
      if (isLogin) {
        const { error } = await supabase.auth.signInWithPassword({
          email,
          password,
        });

        if (error) throw error;

        toast({
          title: "Welcome back!",
          description: "You have successfully signed in.",
        });
        void navigate("/dashboard");
      } else {
        const { error } = await supabase.auth.signUp({
          email,
          password,
          options: {
            data: {
              username: username,
            },
          },
        });

        if (error) throw error;

        toast({
          title: "Account created!",
          description: "Please check your email to verify your account.",
        });
      }
    } catch (error: any) {
      toast({
        title: "Error",
        description: String(error.message),
        variant: "destructive",
      });
    } finally {
      setLoading(false);
    }
  };

  return (
    <div className="min-h-screen bg-nebula flex items-center justify-center px-4">
      <Card className="w-full max-w-md bg-linear-to-br from-brand-800/40 to-slate-800/40 border-brand-500/30 backdrop-blur-xs">
        <CardHeader className="text-center">
          <div className="flex items-center justify-center gap-2 mb-4">
            <Sparkles className="w-8 h-8 text-yellow-400" />
            <span className="text-2xl font-bold text-white">
              Daggerheart Tools
            </span>
          </div>
          <CardTitle className="text-white text-2xl">
            {isLogin ? "Welcome Back" : "Create Account"}
          </CardTitle>
          <CardDescription className="text-brand-200">
            {isLogin
              ? "Sign in to access your characters and campaigns"
              : "Join the adventure and create your first character"}
          </CardDescription>
        </CardHeader>
        <CardContent className="space-y-6">
          <form onSubmit={(e) => void handleAuth(e)} className="space-y-4">
            {/* Honeypot field - hidden from users but visible to bots */}
            <div style={{ display: "none" }}>
              <Label htmlFor="website">Website (leave blank)</Label>
              <Input
                id="website"
                type="text"
                value={honeypot}
                onChange={(e) => {
                  setHoneypot(e.target.value);
                }}
                tabIndex={-1}
                autoComplete="off"
              />
            </div>

            {!isLogin && (
              <div className="space-y-2">
                <Label htmlFor="username" className="text-white">
                  Username
                </Label>
                <Input
                  id="username"
                  type="text"
                  value={username}
                  onChange={(e) => {
                    setUsername(e.target.value);
                  }}
                  className="bg-slate-800/50 border-brand-500/50 text-white"
                  placeholder="Choose a username"
                  required={!isLogin}
                />
              </div>
            )}
            <div className="space-y-2">
              <Label htmlFor="email" className="text-white">
                Email
              </Label>
              <Input
                id="email"
                type="email"
                value={email}
                onChange={(e) => {
                  setEmail(e.target.value);
                }}
                className="bg-slate-800/50 border-brand-500/50 text-white"
                placeholder="Enter your email"
                required
              />
            </div>
            <div className="space-y-2">
              <Label htmlFor="password" className="text-white">
                Password
              </Label>
              <Input
                id="password"
                type="password"
                value={password}
                onChange={(e) => {
                  setPassword(e.target.value);
                }}
                className="bg-slate-800/50 border-brand-500/50 text-white"
                placeholder="Enter your password"
                required
              />
            </div>

            {/* Simple math captcha for registration only */}
            {!isLogin && (
              <div className="space-y-2">
                <Label htmlFor="captcha" className="text-white">
                  Security Check: What is {captchaQuestion.question}?
                </Label>
                <Input
                  id="captcha"
                  type="number"
                  value={captchaAnswer}
                  onChange={(e) => {
                    setCaptchaAnswer(e.target.value);
                  }}
                  className="bg-slate-800/50 border-brand-500/50 text-white"
                  placeholder="Enter the answer"
                  required={!isLogin}
                />
              </div>
            )}

            {/* Cloudflare Turnstile for registration only */}
            {!isLogin && (
              <div className="space-y-2">
                <Label className="text-white">Security Verification</Label>
                <div className="flex justify-center">
                  <Turnstile
                    siteKey="0x4AAAAAABgBiF6BQetCUEj0" // You'll need to replace this with your actual site key
                    onSuccess={(token) => {
                      setTurnstileToken(token);
                    }}
                    onError={() => {
                      setTurnstileToken("");
                    }}
                    onExpire={() => {
                      setTurnstileToken("");
                    }}
                    options={{
                      theme: "dark",
                      size: "normal",
                    }}
                  />
                </div>
              </div>
            )}

            <Button
              type="submit"
              className="w-full bg-yellow-500 hover:bg-yellow-600 text-black font-semibold"
              disabled={loading}
            >
              {loading ? "Loading..." : isLogin ? "Sign In" : "Sign Up"}
            </Button>
          </form>

          <Separator className="bg-brand-500/30" />

          <div className="text-center space-y-4">
            <p className="text-brand-200 text-sm">
              {isLogin ? "Don't have an account?" : "Already have an account?"}
            </p>
            <Button
              variant="outline"
              className="w-full border-brand-400 text-brand-100"
              onClick={() => {
                setIsLogin(!isLogin);
                setCaptchaAnswer(""); // Reset captcha when switching modes
                setTurnstileToken(""); // Reset Turnstile when switching modes
              }}
            >
              {isLogin ? "Create Account" : "Sign In"}
            </Button>
          </div>
        </CardContent>
      </Card>
    </div>
  );
};

export default Auth;
