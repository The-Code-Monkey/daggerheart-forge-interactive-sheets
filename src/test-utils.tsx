import React from "react";
import { render, RenderOptions } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";

// Mock AuthContext
const mockAuthContext = {
  user: null,
  loading: false,
  signIn: jest.fn(),
  signOut: jest.fn(),
  signUp: jest.fn(),
};

const AuthProvider = ({ children }: { children: React.ReactNode }) => {
  return <div data-testid="auth-provider">{children}</div>;
};

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <BrowserRouter>
      <AuthProvider>{children}</AuthProvider>
    </BrowserRouter>
  );
};

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

// Export everything from testing-library/react
export * from "@testing-library/react";
export { customRender as render };

// Mock useAuth hook
export const mockUseAuth = () => mockAuthContext;
