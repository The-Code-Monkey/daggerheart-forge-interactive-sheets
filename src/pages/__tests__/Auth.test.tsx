
import { render } from "../../test-utils";
import Auth from "../Auth";

// Mock react-router-dom
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => jest.fn(),
}));

// Mock supabase
jest.mock("@/integrations/supabase/client", () => ({
  supabase: {
    auth: {
      signInWithPassword: jest.fn(),
      signUp: jest.fn(),
    },
  },
}));

describe("Auth Page", () => {
  it("renders correctly", () => {
    const { container } = render(<Auth />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
