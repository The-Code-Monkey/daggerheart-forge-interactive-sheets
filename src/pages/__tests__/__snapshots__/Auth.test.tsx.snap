// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth Page renders correctly 1`] = `
<div
  data-testid="auth-provider"
>
  <div
    class="min-h-screen bg-linear-to-b from-brand-900 via-brand-800 to-slate-900 flex items-center justify-center px-4"
  >
    <div
      class="rounded-lg border text-card-foreground shadow-sm bg-gradient-to-br w-full max-w-md bg-linear-to-br from-brand-800/40 to-slate-800/40 border-brand-500/30 backdrop-blur-xs"
    >
      <div
        class="flex flex-col space-y-1.5 p-6 text-center"
      >
        <div
          class="flex items-center justify-center gap-2 mb-4"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-sparkles w-8 h-8 text-yellow-400"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9.937 15.5A2 2 0 0 0 8.5 14.063l-6.135-1.582a.5.5 0 0 1 0-.962L8.5 9.936A2 2 0 0 0 9.937 8.5l1.582-6.135a.5.5 0 0 1 .963 0L14.063 8.5A2 2 0 0 0 15.5 9.937l6.135 1.581a.5.5 0 0 1 0 .964L15.5 14.063a2 2 0 0 0-1.437 1.437l-1.582 6.135a.5.5 0 0 1-.963 0z"
            />
            <path
              d="M20 3v4"
            />
            <path
              d="M22 5h-4"
            />
            <path
              d="M4 17v2"
            />
            <path
              d="M5 18H3"
            />
          </svg>
          <span
            class="text-2xl font-bold text-white"
          >
            Daggerheart Tools
          </span>
        </div>
        <h3
          class="font-semibold tracking-tight text-white text-2xl"
        >
          Welcome Back
        </h3>
        <p
          class="text-sm text-brand-200"
        >
          Sign in to access your characters and campaigns
        </p>
      </div>
      <div
        class="p-6 pt-0 space-y-6"
      >
        <form
          class="space-y-4"
        >
          <div
            style="display: none;"
          >
            <label
              class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
              for="website"
            >
              Website (leave blank)
            </label>
            <input
              autocomplete="off"
              class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-base ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-hidden disabled:cursor-not-allowed disabled:opacity-50 focus-visible:border-2 focus-visible:border-white md:text-sm"
              id="website"
              tabindex="-1"
              type="text"
              value=""
            />
          </div>
          <div
            class="space-y-2"
          >
            <label
              class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 text-white"
              for="email"
            >
              Email
            </label>
            <input
              class="flex h-10 w-full rounded-md border px-3 py-2 text-base ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-hidden disabled:cursor-not-allowed disabled:opacity-50 focus-visible:border-2 focus-visible:border-white md:text-sm bg-slate-800/50 border-brand-500/50 text-white"
              id="email"
              placeholder="Enter your email"
              required=""
              type="email"
              value=""
            />
          </div>
          <div
            class="space-y-2"
          >
            <label
              class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 text-white"
              for="password"
            >
              Password
            </label>
            <input
              class="flex h-10 w-full rounded-md border px-3 py-2 text-base ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-hidden disabled:cursor-not-allowed disabled:opacity-50 focus-visible:border-2 focus-visible:border-white md:text-sm bg-slate-800/50 border-brand-500/50 text-white"
              id="password"
              placeholder="Enter your password"
              required=""
              type="password"
              value=""
            />
          </div>
          <button
            class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm ring-offset-background transition-colors duration-300 ease-in-out focus-visible:outline-hidden focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 h-10 px-4 py-2 w-full bg-yellow-500 hover:bg-yellow-600 text-black font-semibold"
            type="submit"
          >
            Sign In
          </button>
        </form>
        <div
          class="shrink-0 h-px w-full bg-brand-500/30"
          data-orientation="horizontal"
          role="none"
        />
        <div
          class="text-center space-y-4"
        >
          <p
            class="text-brand-200 text-sm"
          >
            Don't have an account?
          </p>
          <button
            class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors duration-300 ease-in-out focus-visible:outline-hidden focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 border bg-brand-700/80 hover:bg-brand-700 hover:text-accent-foreground h-10 px-4 py-2 w-full border-brand-400 text-brand-100"
          >
            Create Account
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
